File: mpbeps3/ReadFiles/pflreadf3.py
#-----------------------------------------------------------------------
# This program reads real periodic 3d fluid data
# written by 3D MPI/OpenMP PIC codes
# written by Viktor K. Decyk, UCLA
from __future__ import print_function
import sys
import math
import numpy

sys.path.append('./ReadFiles')
from libcmfield3 import *

if (sys.version_info.major==2):
   input = raw_input

#-----------------------------------------------------------------------
def main(idrun):
   """
   reads real periodic 3d fluid data
   idrun = run identifier for current run, -1 if unknown
   """
   int_type = numpy.int32
   double_type = numpy.float64
   float_type = numpy.float32
   complex_type = numpy.complex64

   ns = 2
   iudm = 19; iuv = 12
   dname = numpy.array(["ELECTRON FLUID MOMENTS","ION FLUID MOMENTS     "],
                       dtype=str)
   sname = numpy.array(["ELECTRON","ION     "],dtype=str)
   ename = numpy.array([" DENSITY        "," VELOCITY FIELD ",
                        " PRESSURE TENSOR"," ENERGY         ",
                        " HEAT FLUX      "],dtype=str)

# create string from idrun
   if (idrun < 0):
      cdrun = "Unknown"
      while (cdrun.isdigit() == False):
         cdrun = input("enter integer idrun: ")
      idrun = int(cdrun)
   cdrun = str(idrun)
   fname = "diag3." + cdrun
   cmfield3.ffopen3(iudm,fname)

# nscalars = table of available diagnostics
   nscalars = numpy.zeros((ns),int_type,'F')

# determine which fluid diagnostics are available
   cmfield3.readfldiags3(iudm,nscalars)

   nts = numpy.zeros((1),int_type,'F')
   npro = numpy.zeros((1),int_type,'F')
   nprd = numpy.zeros((1),int_type,'F')
   mrec = numpy.zeros((1),int_type,'F')
   fname = numpy.array([""],'S32')

# select diagnostic
   m = numpy.sum(nscalars)
   while True:
      if (m > 0):
         n = -1
         while True:
            if (n < 0):
               for i in range(0,ns):
                  if (nscalars[i]==1):
                     print ("enter ", i+1," for", 
                            numpy.str.rstrip(dname[i]))
               print ("enter ", 0," for EXIT")
               c = input("")
               if (c.isdigit()):
                  n = int(c)
               if (n==0):
                  break
               if ((n >= 1) and (n <= ns)):
                  if (nscalars[n-1]==0):
                     n = -1
            else:
               n = -1
            if (n > 0):
               break
            print ("invalid entry, try again or enter 0 to quit")
      else:
         print ("no fluid diagnostic files found")
         n = 0
# exit procedure
      if (n==0):
         if ("sfield" in globals()):
            sfield = None
         if ("fms" in globals()):
            sfield = None
         if ("vfield" in globals()):
            ufield = None
         if ("vfield" in globals()):
            ufield = None
         cmfield3.closeff3(iudm)
         return

      print (numpy.str.rstrip(dname[n-1]), " diagnostic selected")

# return parameters for selected fluid diagnostic:
# nts, npro, nprd, nrec, fname
      cmfield3.fldiagparams3(iudm,n,nts,npro,nprd,mrec,fname)
      nrec = mrec[0]

# nx/ny/nz = number of global grid points in x/y/z direction
      nx = int(math.pow(2,in3.indx)); ny = int(math.pow(2,in3.indy))
      nz = int(math.pow(2,in3.indz))
# kyp/kzp = number of real grids in each field partition in y/z
      kyp = int((ny - 1)/in3.nvpy) + 1; kzp = int((nz - 1)/in3.nvpz) + 1
# kyb/kzb = minimum number of processors in distributed array in y/z
      kyb = int((ny - 1)/kyp) + 1; kzb = int((nz - 1)/kzp) + 1
# nyv = second dimension of scalar field array, >= ny
# nzv = third dimension of scalar field array, >= nz
      nyv = kyp*kyb; nzv = kzp*kzb

# allocate vector array
      if ("fms" not in globals()):
         fms = numpy.empty((nprd[0],nx,nyv,nzv),float_type,'F')
      if ("sfield" not in globals()):
         sfield = numpy.empty((nx,nyv,nzv),float_type,'F')
      if (in3.ndim==3):
         if ("vfield" not in globals()):
            vfield = numpy.empty((in3.ndim,nx,nyv,nzv),float_type,'F')
         if (npro[0] > 2):
            if ("ufield" not in globals()):
               ufield = numpy.empty((2*in3.ndim,nx,nyv,nzv),float_type,'F')
      dt = in3.dt*float(nts[0])

# open stream file for vector field
      cmfield3.fsopen3(iuv,fname)

# nrec = number of complete records
      nrec = int(nrec/(kyb*kzb))
      print ("records found: nrec = ", nrec)

# read and transpose vector data
      for ii in range(0,nrec):
# read real vector field
         cmfield3.freadv3(iuv,fms,nprd,nx,kyp,kyb,kzp,kzb)
         it = nts[0]*ii
         time = dt*float(ii)
# show time
         print (sname[n-1], " it,time=",it,time)
      cmfield3.closeff3(iuv)
      print()

if (__name__=="__main__"):
   main(-1)

File: mpbeps3/ReadFiles/Python2/pflreadf3.f90
!-----------------------------------------------------------------------
! This program reads real periodic 3d fluid data
! written by 3D MPI/OpenMP PIC codes
! written by Viktor K. Decyk, UCLA
      program preadf3
      use in3, only: idrun, indx, indy, indz, nvpy, nvpz, ndim, tend, dt
      use cmfield3
!
      implicit none
      integer, parameter :: ns = 2
      integer :: iudm = 19, iuv = 12
      integer :: i, m, n, ii, it, nx, ny, nz, kyp, kzp, kyb, kzb
      integer :: nyv, nzv, npro, nprd, nrec
      integer :: nts = 0
      real :: time
! nscalars = table of available diagnostics
      integer, dimension(ns) :: nscalars = 0
      real, dimension(:,:,:,:), allocatable :: fms, vfield, ufield
      real, dimension(:,:,:), allocatable :: sfield
      character(len=20), dimension(ns) :: dname = (/                    &
     &'elect fluid moments ','ion fluid moments   '/)
      character(len=8), dimension(ns) :: sname = (/'ELECTRON','ION     '&
     &/)
      character(len=16), dimension(5) :: ename = (/' DENSITY        ',  &
     &' VELOCITY FIELD ',' PRESSURE TENSOR',' ENERGY         ',         &
     &' HEAT FLUX      '/)
      character(len=10) :: cdrun
      character(len=32) :: fname
!
      write (*,*) 'enter idrun:'
      read (5,*) idrun
      write (cdrun,'(i10)') idrun
      cdrun = adjustl(cdrun)
      fname = 'diag3.'//cdrun
      call ffopen3(iudm,fname)
!
! determine which fluid diagnostics are available
      call readfldiags3(iudm,nscalars)
!
! select diagnostic
      m = sum(nscalars)
      if (m > 1) then
         n = -1
         do while (n < 0)
            do i = 1, ns
               if (nscalars(i)==1) then
                  write (*,*) 'enter ', i, 'for ', trim(dname(i))
               endif
            enddo
            read (5,*) n
            if (n==0) stop
            if ((n >= 1).and.(n <= ns)) then
               if (nscalars(n)==0) n = -1
            else
               n = -1
            endif
            if (n < 0) then
               write (*,*) 'invalid entry, try again or enter 0 to quit'
            endif
         enddo
      else if (m==1) then
         do i = 1, ns
            if (nscalars(i)==1) then
               n = i
               exit
            endif
         enddo
      else
         write (*,*) 'no fluid moments diagnostic files found'
         stop
      endif
!
      write (*,*) trim(dname(n)), ' diagnostic selected'
!
! return parameters for selected fluid diagnostic
      call fldiagparams3(iudm,n,nts,npro,nprd,nrec,fname)
!
! nx/ny/nz = number of global grid points in x/y/z direction
      nx = 2**indx; ny = 2**indy; nz = 2**indz
! kyp/kzp = number of real grids in each field partition in y/z
      kyp = (ny - 1)/nvpy + 1; kzp = (nz - 1)/nvpz + 1
! kyb/kzb = minimum number of processors in distributed array in y/z
      kyb = (ny - 1)/kyp + 1; kzb = (nz - 1)/kzp + 1
! nyv = second dimension of scalar field array, >= ny
! nzv = third dimension of scalar field array, >= nz
      nyv = kyp*kyb; nzv = kzp*kzb
!
! allocate vector array
      allocate(fms(nprd,nx,nyv,nzv),sfield(nx,nyv,nzv))
      if (ndim==3) then
         allocate(vfield(ndim,nx,nyv,nzv))
         if (npro > 2) allocate(ufield(2*ndim,nx,nyv,nzv))
      endif
      dt = dt*real(nts)
!
! open stream file for vector field
      call fsopen3(iuv,fname)
!
! nrec = number of complete records
      nrec = nrec/(kyb*kzb)
      write (*,*) 'records found: nrec = ', nrec
!
! read and transpose vector data
      do ii = 1, nrec
! read real vector field
         call freadv3(iuv,fms,nprd,nx,kyp,kyb,kzp,kzb)
         it = nts*(ii - 1)
         time = dt*real(ii - 1)
! show time
         write (*,*) sname(n),'it,time=',it,time
      enddo
!
      call closeff3(iudm)
      call closeff3(iuv)
!
      end program
